src/Http/Http.Results/src/FileContentHttpResult.cs

```diff
@@ -0,0 +1,124 @@
+    /// <param name="fileDownloadName">The suggested file name.</param>
+    /// <param name="enableRangeProcessing">Set to <c>true</c> to enable range requests processing.</param>
+    /// <param name="lastModified">The <see cref="DateTimeOffset"/> of when the file was last modified.</param>
+    /// <param name="entityTag">The <see cref="EntityTagHeaderValue"/> associated with the file.</param>
+    internal FileContentHttpResult(
+        ReadOnlyMemory<byte> fileContents,
+        string? contentType,
+        string? fileDownloadName,
+        bool enableRangeProcessing,
+        DateTimeOffset? lastModified = null,
+        EntityTagHeaderValue? entityTag = null)
+    {
+        FileContents = fileContents;
+        FileLength = fileContents.Length;
+        ContentType = contentType ?? "application/octet-stream";
+        FileDownloadName = fileDownloadName;
+        EnableRangeProcessing = enableRangeProcessing;
+        LastModified = lastModified;
+        EntityTag = entityTag;
+    }
+
+    /// <summary>
+    /// Gets the Content-Type header for the response.
+    /// </summary>
+    public string ContentType { get; internal set; }
+
+    /// <summary>
+    /// Gets the file name that will be used in the Content-Disposition header of the response.
+    /// </summary>
+    public string? FileDownloadName { get; internal set; }
+
+    /// <summary>
+    /// Gets the last modified information associated with the file result.
+    /// </summary>
+    public DateTimeOffset? LastModified { get; internal set; }
+
+    /// <summary>
+    /// Gets the etag associated with the file result.
+    /// </summary>
+    public EntityTagHeaderValue? EntityTag { get; internal init; }
+
+    /// <summary>
+    /// Gets the value that enables range processing for the file result.
+    /// </summary>
+    public bool EnableRangeProcessing { get; internal init; }
+
+    /// <summary>
+    /// Gets or sets the file length information .
+    /// </summary>
+    public long? FileLength { get; internal set; }
+
+    /// <summary>
+    /// Gets or sets the file contents.
+    /// </summary>
+    public ReadOnlyMemory<byte> FileContents { get; internal init; }
+
+    /// <inheritdoc/>
+    public Task ExecuteAsync(HttpContext httpContext)
+    {
+        // Creating the logger with a string to preserve the category after the refactoring.
+        var loggerFactory = httpContext.RequestServices.GetRequiredService<ILoggerFactory>();
+        var logger = loggerFactory.CreateLogger("Microsoft.AspNetCore.Http.Result.FileContentResult");
+
+        return HttpResultsHelper.WriteResultAsFileAsync(
+            httpContext,
+            logger,
+            FileDownloadName,
+            FileLength,
+            ContentType,
+            EnableRangeProcessing,
+            LastModified,
+            EntityTag,
+            (context, range, rangeLength) => FileResultHelper.WriteFileAsync(httpContext, FileContents, range, rangeLength));
```

Comment:
Could avoid the closure alloc by passing FileContents into WriteResultAsFileAsync and passing it in the Func